File: 二叉树/前根遍历.js
const preorderTraversal = root => {
  const result = []
  const stack = []
  let T = root
  T && stack.push(T)
  while (stack.length) {
    T = stack.pop()
    result.push(T.val)
    while (T) {
      if (T.lchild) {
        result.push(T,lchild.val)
      }
      if (T.rchild) {
        stack.push(T.rchild)
      }
      T = T.lchild
    }
  }

  return result 
}

File: 二叉树/前根遍历.java
public int[] preorderTraversal(root) {
  BiTreeNode T = root;
  int[] result = new int[];
  if (T != null) {
    LinkStack S = new LinkStack();
    S.push(T);
    while (!S.isEmpty()) {
      T = (BiTreeNode)S.pop();
      result.push(T.val);
      while (T != null) {
        if (T.lchild != null) {
          result.push(T.lchild.val);
        }
        if (T.rchild != null) {
          S.push(T.rchild);
        }
        T = T.lchild;
      }
    }
  }
}
